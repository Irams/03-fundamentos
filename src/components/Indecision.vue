<template>
    <img v-if="img" :src="img" alt="bg">
    <div class="bg-dark"></div>
    <div class="indecision-container">
        <input type="text"
               v-model="question"
               >
        <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>

        <div v-if="validQuestion">
            <h2>{{question}}</h2>
            <h1 >{{answer}}</h1>
        </div>
    </div>
  
</template>

<script>
export default {
    data(){
       return{
           question: null,
           answer: null,
           img: null,
           validQuestion: false,
           translateAnswer: null
       } 
    },
    methods: {
        async getAnswer(){

            try {
                this.answer = 'Pensando...'
    
                const {answer, image} = await fetch('https://yesno.wtf/api').then(resp => resp.json())
     
                this.answer = answer === 'yes' ? 'Si!' : 'No!'
                this.img = image                
            } catch (error) {
                console.log('Indecision Component: ', error);
                this.answer = 'No se pudo cargar desde el API'
                this.img = null
            }
        },
    },
    watch: {
        question(value, oldValue){

            this.validQuestion = false

            console.log({value})

            if(!value.includes('?')) return

            this.validQuestion = true
            console.log({value})
            
            //TODO: realizar peticion http
            this.getAnswer()

        }
    }
}
</script>

//el scope aplica estos estilos solo a este componente
<style scoped>

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 15px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>