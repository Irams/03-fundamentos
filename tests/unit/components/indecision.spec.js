import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision'

describe('Indecision Component', ()=>{
    let wrapper
    let himsSpy

    //Mock del fetch:
    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }))

    beforeEach(() =>{
        wrapper = shallowMount(Indecision)

        himsSpy = jest.spyOn(console, 'log')

        jest.clearAllMocks()

    })

    test('match con el snapshot', ()=>{
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('escribir en el input no debe dispara nada (console.log)', async ()=>{

        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('Hola Mundo')

        expect(himsSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()

        // console.log(wrapper.vm);

    })

    test('escribir el simbolo de "?" debe disparar el getAnswer', async ()=>{
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
        const input = wrapper.find('input')
        await input.setValue('Va a llover?')
        expect(himsSpy).toHaveBeenCalledTimes(2)
        expect(getAnswerSpy).toHaveBeenCalled()

    })

    test('pruebas en getAnswer', async ()=>{
    
        await wrapper.vm.getAnswer()

        // console.log(wrapper.vm.img)
        // console.log(wrapper.vm.answer)

        const img = wrapper.find('img')

        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Si!')

    })

    test('pruebas en getAnswer - fallo en el API', async() => {

        fetch.mockImplementationOnce(() => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')
        
        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('No se pudo cargar desde el API')
    })
})