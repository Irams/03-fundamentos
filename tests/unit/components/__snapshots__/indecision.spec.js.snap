// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text">
  <p>Recuerda terminar con un signo de interrogación (?)</p>
  <!--v-if-->
</div>
`;

exports[`Indecision Component match con el snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text">
  <p>Recuerda terminar con un signo de interrogación (?)</p>
  <!--v-if-->
</div>
`;
